File: ios/ControlFrame.swift
import Foundation

/// Typed control-socket frame from the Node backend. Mirrors the Android
/// `ControlFrame` sealed class.
///
/// Adding a new frame type means a case here AND a branch on each
/// consumer's switch — Swift exhaustiveness is the point.
enum ControlFrame {
    case started
    case ready
    /// Backend has begun graceful shutdown. A control-socket close
    /// without a preceding `.stopping` is unambiguously a crash or kill.
    case stopping
    case error(phase: String, message: String)
    /// Sentry error event from `@sentry/node`, forwarded as JSON for
    /// `SentryEventDecoder.decodeEvent(jsonData:)`. Native scope (device,
    /// OS, app, user) merges so Node doesn't have to carry it.
    case sentryEvent(payloadJson: String)
    /// Sentry envelope from `@sentry/node` (transactions, sessions,
    /// check-ins, profiles). Base64 bytes handed to
    /// `PrivateSentrySDKOnly`. No native scope merging.
    case sentryEnvelope(data: String)
    /// BLE engine commands from the backend's discovery controller
    /// (`backend/lib/ble-discovery.js`). Dispatched to
    /// `BleDiscoveryEngine` via `NodeJSService.onBleControlFrame`.
    /// `payload` is the base64 v1 advertisement (nil = scan-only /
    /// stop advertising).
    case bleStart(payload: String?)
    case bleAdvertise(payload: String?)
    case bleStop
    /// DNS-SD engine commands: register `_comapeo._tcp` as name:port and
    /// browse for peers (`NsdEngine`), or stop.
    case nsdStart(name: String, port: Int)
    case nsdStop
    /// Not JSON, missing `type`, or `type` not in the well-known set.
    /// `detail` is developer-facing — surfaces in the JS `messageerror`.
    case malformed(detail: String)

    /// Never throws; every failure resolves to `.malformed`.
    static func parse(_ raw: String) -> ControlFrame {
        guard let data = raw.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            return .malformed(detail: "Non-JSON control frame: \(raw.prefix(100))")
        }
        let type = (obj["type"] as? String) ?? ""
        switch type {
        case "started":
            return .started
        case "ready":
            return .ready
        case "stopping":
            return .stopping
        case "error":
            let phase = (obj["phase"] as? String) ?? "unknown"
            let message = (obj["message"] as? String) ?? "(no message)"
            return .error(phase: phase, message: message)
        case "sentry-event":
            guard let payload = obj["payload"] as? [String: Any] else {
                return .malformed(detail: "sentry-event frame missing object `payload`")
            }
            // Re-stringify so the decoder can re-parse from bytes.
            guard let bytes = try? JSONSerialization.data(withJSONObject: payload, options: []),
                  let json = String(data: bytes, encoding: .utf8)
            else {
                return .malformed(detail: "sentry-event frame `payload` is not serializable JSON")
            }
            return .sentryEvent(payloadJson: json)
        case "sentry-envelope":
            guard let data = obj["data"] as? String else {
                return .malformed(detail: "sentry-envelope frame missing string `data`")
            }
            return .sentryEnvelope(data: data)
        case "ble-start":
            return .bleStart(payload: obj["payload"] as? String)
        case "ble-advertise":
            return .bleAdvertise(payload: obj["payload"] as? String)
        case "ble-stop":
            return .bleStop
        case "nsd-start":
            guard let name = obj["name"] as? String, !name.isEmpty,
                  let port = obj["port"] as? Int, port > 0
            else {
                return .malformed(detail: "nsd-start frame missing `name`/`port`")
            }
            return .nsdStart(name: name, port: port)
        case "nsd-stop":
            return .nsdStop
        default:
            return .malformed(detail: "Unknown control frame type=\"\(type)\"")
        }
    }
}

File: android/src/main/java/com/comapeo/core/ControlFrame.kt
package com.comapeo.core

import org.json.JSONException
import org.json.JSONObject

/**
 * Typed control-socket frame from the embedded Node.js backend. Names match what
 * `backend/lib/simple-rpc.js` emits / `backend/index.js` broadcasts. The compiler's
 * exhaustiveness check on consumers is the point — adding a frame fails the build
 * until every `when` is updated, rather than silently dropping it at runtime.
 */
sealed class ControlFrame {
    object Started : ControlFrame()
    object Ready : ControlFrame()

    /** Graceful shutdown — sent before close so peers can tell expected from crash. */
    object Stopping : ControlFrame()

    data class Error(val phase: String, val message: String) : ControlFrame()

    /**
     * `@sentry/node` error event, JSON-encoded. Fed to `SentryEvent.Deserializer`
     * + `Sentry.captureEvent` so native scope (device/OS/app/user) is merged at
     * capture time and Node doesn't have to carry it.
     */
    data class SentryEvent(val payloadJson: String) : ControlFrame()

    /**
     * `@sentry/node` envelope (transactions, sessions, check-ins, profiles, …)
     * base64-encoded. Handed to the hybrid-SDK envelope entrypoint for offline-
     * capable transport. Native scope is NOT applied — parent transaction is
     * opened natively and Node spans inherit via `continueTrace`.
     */
    data class SentryEnvelope(val data: String) : ControlFrame()

    /**
     * BLE engine commands from the backend's discovery controller
     * (`backend/lib/ble-discovery.js`), which owns the whole discovery
     * lifecycle. Broadcast frames — the FGS's [NodeJSService] dispatches
     * them to the engine; every other control client ignores them.
     * `payload` is the base64 v1 advertisement (null = scan-only /
     * stop advertising).
     */
    data class BleStart(val payload: String?) : ControlFrame()
    data class BleAdvertise(val payload: String?) : ControlFrame()
    object BleStop : ControlFrame()

    /**
     * DNS-SD engine commands, same contract as the BLE ones: the
     * backend's discovery controller tells the FGS-hosted [NsdEngine]
     * to register `_comapeo._tcp` as `name`:`port` and browse for
     * peers, or to stop.
     */
    data class NsdStart(val name: String, val port: Int) : ControlFrame()
    object NsdStop : ControlFrame()

    /** Frame could not be processed; `detail` is suitable for logs / `messageerror`. */
    data class Malformed(val detail: String) : ControlFrame()

    companion object {
        /** Never throws; every failure mode resolves to [Malformed]. */
        fun parse(raw: String): ControlFrame {
            val json = try {
                JSONObject(raw)
            } catch (_: JSONException) {
                return Malformed("Non-JSON control frame: ${raw.take(100)}")
            }
            return when (val type = json.optString("type", "")) {
                "started" -> Started
                "ready" -> Ready
                "stopping" -> Stopping
                "error" -> Error(
                    phase = json.optString("phase", "unknown"),
                    message = json.optString("message", "(no message)"),
                )
                "sentry-event" -> {
                    val payload = json.optJSONObject("payload")
                    // Re-serialize so SentryEvent.Deserializer can re-parse against the bytes it expects.
                    payload?.let { SentryEvent(it.toString()) }
                        ?: Malformed("sentry-event frame missing object `payload`")
                }
                "sentry-envelope" -> {
                    val data = json.optString("data", "")
                    if (data.isEmpty()) Malformed("sentry-envelope frame missing string `data`")
                    else SentryEnvelope(data)
                }
                "ble-start" -> BleStart(payload = json.optStringOrNull("payload"))
                "ble-advertise" -> BleAdvertise(payload = json.optStringOrNull("payload"))
                "ble-stop" -> BleStop
                "nsd-start" -> {
                    val name = json.optString("name", "")
                    val port = json.optInt("port", 0)
                    if (name.isEmpty() || port <= 0) {
                        Malformed("nsd-start frame missing `name`/`port`")
                    } else {
                        NsdStart(name = name, port = port)
                    }
                }
                "nsd-stop" -> NsdStop
                else -> Malformed("Unknown control frame type=\"$type\"")
            }
        }

        /** `null` for an absent key or a JSON `null` value. */
        private fun JSONObject.optStringOrNull(key: String): String? =
            if (isNull(key)) null else optString(key)
    }
}
